# Spam Detector

In [20]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [21]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [22]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]

# Create the features DataFrame (X) from the remaining columns
X = data.drop(columns=["spam"])

# Display the shapes of the created sets
X.shape, y.shape

((4601, 57), (4601,))

In [23]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
label_balance = y.value_counts()
label_balance

spam
0    2788
1    1813
Name: count, dtype: int64

In [24]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1, stratify=y)

# Display the shapes of the split datasets
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((3680, 57), (921, 57), (3680,), (921,))

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [25]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [26]:
# Fit the Standard Scaler with the training data
# Fit the scaler on the training data and transform both training and testing sets
X_train_scaled = scaler.fit_transform(X_train)


# Display the first few rows of the scaled training data
pd.DataFrame(X_train_scaled).head()

0         1         2         3         4         5         6   \
0 -0.349630 -0.164243 -0.555073 -0.046305  0.357965 -0.356591 -0.297159   
1 -0.349630 -0.164243 -0.555073 -0.046305 -0.457379 -0.356591 -0.297159   
2 -0.349630 -0.164243 -0.555073  0.161749  0.887223 -0.356591 -0.297159   
3  0.097203 -0.063116 -0.292817 -0.046305  0.329356  0.632379  0.400139   
4 -0.040284 -0.094232 -0.373511 -0.046305 -0.042555 -0.356591 -0.297159   

         7         8         9   ...        47        48        49        50  \
0 -0.255351 -0.333617 -0.360622  ... -0.113219 -0.158144 -0.220171 -0.171164   
1 -0.255351 -0.333617 -0.360622  ... -0.113219 -0.158144 -0.489416 -0.171164   
2  0.483831 -0.333617  0.601768  ... -0.113219 -0.158144 -0.360038 -0.171164   
3  0.054629  3.876585  0.051831  ... -0.113219 -0.158144 -0.426475 -0.171164   
4 -0.255351  2.966271 -0.360622  ... -0.113219  0.288174  0.045578  0.538650   

         51        52        53        54        55        56  
0 -0.309557 -0.314541 -0.099409 -0.101496 -0.194021 -0.280121  
1  0.423797 -0.314541 -0.099409 -0.128847 -0.246180 -0.453794  
2 -0.223689 -0.314541 -0.099409 -0.044974 -0.065994 -0.061000  
3  1.191963  0.427929 -0.099409 -0.020684  0.849162  0.718094  
4 -0.309557 -0.147281 -0.099409 -0.062477 -0.056510  0.653169  

[5 rows x 57 columns]

In [27]:
# Scale the training data
X_test_scaled = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [28]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
logistic_regression_model = LogisticRegression(random_state=1)
# Fit the model to the scaled training data
logistic_regression_model.fit(X_train_scaled, y_train)

# Make predictions using the scaled testing data
y_pred_logistic = logistic_regression_model.predict(X_test_scaled)

In [29]:
# Make and save testing predictions with the saved logistic regression model using the test data
logistic_predictions = pd.DataFrame({
    "Actual": y_test,
    "Predicted": y_pred_logistic
})
# Review the predictions
logistic_predictions.head()

Actual  Predicted
1039       1          1
3312       0          1
3101       0          0
2066       0          0
4299       0          0

In [30]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
# Calculate the accuracy score by comparing the actual and predicted values
accuracy = accuracy_score(logistic_predictions["Actual"], logistic_predictions["Predicted"])

# Display the calculated accuracy
accuracy

0.9337676438653637

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [31]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier


In [32]:
# Make and save testing predictions with the saved logistic regression model using the test data


# Review the predictions

random_forest_model = RandomForestClassifier(random_state=1)

# Fit the model to the scaled training data
random_forest_model.fit(X_train_scaled, y_train)

# Make predictions using the scaled testing data
y_pred_forest = random_forest_model.predict(X_test_scaled)




In [33]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
# Calculate and print the accuracy score of the Random Forest model
forest_accuracy = accuracy_score(y_test, y_pred_forest)
forest_accuracy

0.9511400651465798

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Replace the text in this markdown cell with your answers to these questions.*

# Logistic Regression Model Accuracy: 93.38%
# Random Forest Classifier Accuracy: 95.11%
As predicted, the Random Forest Classifier outperformed the Logistic Regression model. The Random Forest model, being an ensemble method, is generally more powerful for classification tasks like this, as it can capture complex patterns and interactions between features. This makes it more robust than a linear model like Logistic Regression, especially when the dataset contains various features that may not have linear relationships with the target variable.

Summary of Findings:
Random Forest outperformed Logistic Regression by about 1.73% in accuracy.
The initial prediction that Random Forest would perform better was correct, as it tends to handle more complex data better by leveraging multiple decision trees and aggregating their results.